### PR TITLE
Have all numeric OSSL_PARAM_get_xxx() check that data and data_size are set

### PR DESCRIPTION
**Pull request by levitte**
If the OSSL_PARAM fields |data| or |data_size| are not set (data == NULL
or data_size == 0), it's not a valid number.

### CODE REVIEW

**Comment by ifranzki** [comment]
My comment from the other PR still applies:

You added the check to OSSL_PARAM_get_int32() etc, but not to OSSL_PARAM_get_int() etc. Within OSSL_PARAM_get_int() it calls general_get_int(). So no check before it reaches general_get_int().

This applies to OSSL_PARAM_get_int(), OSSL_PARAM_get_long(), OSSL_PARAM_get_time_t(), OSSL_PARAM_get_uint(), OSSL_PARAM_get_size_t(), OSSL_PARAM_get_ulong().

**Comment by levitte** [comment]
Thanks!  I glossed over the end of those functions, it seems.  Fixup commit coming up!

**Comment by levitte** [comment]
As soon as #20013 is in and I rebase this, CIs should perform better, probably except for "Provider old versions compat"

**Comment by ifranzki** [comment]
> Thanks! I glossed over the end of those functions, it seems. Fixup commit coming up!

OK, you could have added the check just into `general_get_[u]int()` instead of each individual function, but that's up to you....

Regarding the "Provider old versions compat" failures: Is this happening just for zero BNs that are passed as `data = NULL` and `data_size = 0`? If so, then maybe you could allow them being NULL for `OSSL_PARAM_get_BN()` and special-case this case to create a zero BN right away in` OSSL_PARAM_get_BN()` without going to `BN_native2bn()`....

Remember, calling `BN_native2bn()` with `data = NULL` and `data_size = 0` causes a address sanitizer error they  way the code is currently. We should avoid that, even if we allow `data = NULL` and `data_size = 0` for this getter.

File: crypto/params.c [hunk 1]
```diff
@@ -258,6 +260,15 @@ int OSSL_PARAM_get_int(const OSSL_PARAM *p, int *val)
         return OSSL_PARAM_get_int64(p, (int64_t *)val);
     }
 #endif
+
```
Comment (paulidale) [inline crypto/params.c:263]: This should go before the `#ifndef OPENSSL_SMALL_FOOTPRINT` block.  Both here and everywhere else that has such.

**Review by paulidale — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
> OK, you could have added the check just into `general_get_[u]int()` instead of each individual function, but that's up to you....

This doesn't work.  There are fast paths when building without the small memory define that don't go through the `general_` functions.

**Comment by ifranzki** [comment]
> > OK, you could have added the check just into `general_get_[u]int()` instead of each individual function, but that's up to you....
> 
> This doesn't work. There are fast paths when building without the small memory define that don't go through the `general_` functions.

Please note the first commit in this PR. This adds the checks to exactly those fastpath functions. So those are already covered.

**Comment by paulidale** [comment]
The checks are added **after** the fasts paths have returned.

**Comment by ifranzki** [comment]
> The checks are added **after** the fasts paths have returned.

Well, if it goes the fastpath way, it will never reach the `general_get_int()` call, because the switch cases return from the function:

    switch (sizeof(int)) {
    case sizeof(int32_t):
        return OSSL_PARAM_get_int32(p, (int32_t *)val);
    case sizeof(int64_t):
        return OSSL_PARAM_get_int64(p, (int64_t *)val);
    }

So the `general_get_int()` call can only be reached if `OPENSSL_SMALL_FOOTPRINT` is defined. And then it should be sufficient to add the check into the `general_get_int()` function (as I suggested), or before the call to `general_get_int()` (as it is done in the 2nd commit), no?

**Comment by paulidale** [comment]
The general calls can still be reached.  E.g. I get a single byte number and want to fit it into a 4 byte integer -> the general path is followed.

**Comment by ifranzki** [comment]
> The general calls can still be reached. E.g. I get a single byte number and want to fit it into a 4 byte integer -> the general path is followed.

I don't see how.... The switch is for `sizeof(int)` and this is constant at runtime. Its NOT a switch for the data length passed in the OSSL_PARAM! 
So if OPENSSL_SMALL_FOOTPRINT is not defined it will never go to the general call. 
But even if it would, adding the check into the `general_get_[u]int()` functions (or before the call to them) would still be sufficient.

**Comment by paulidale** [comment]
Look in the int32 function.  If the data size isn't 4 or 8 bytes, it falls through to the general function.

The params convert types.  If I have a one byte bignum that is passed as a one byte value (e.g. a zero now), the param code will convert this to the desired size.  For sizes it doesn't know about (i.e. anything except 4 or 8 bytes), the general code is called.

**Comment by ifranzki** [comment]
> Look in the int32 function. If the data size isn't 4 or 8 bytes, it falls through to the general function.

Right, but look at the first commit (https://github.com/openssl/openssl/pull/20017/commits/6750731ba98b4b0c0b5ebab9549d85fedf4f4924) where the check is added to OSSL_PARAM_get_int32 (and the others).
The check is right at the beginning, so it is checked before the general function is called.

```
int OSSL_PARAM_get_int32(const OSSL_PARAM *p, int32_t *val)
{
    double d;
    if (val == NULL || p == NULL) {
        err_null_argument;
        return 0;
    }
    if (p->data_size == 0 ||  p->data == NULL) {     <---- see here
        err_invalid_argument;
        return 0;
    }
    ....
```

So I am still convinced that the way the code is currently (both commits!) is sufficient.

File: crypto/params.c (hunk 1 above)
Comment (levitte) [reply crypto/params.c:263]: No, that would be wrong

Comment (levitte) [reply crypto/params.c:263]: Small footprint doesn't mean no checks for invalid input

**Comment by levitte** [comment]
> > The general calls can still be reached. E.g. I get a single byte number and want to fit it into a 4 byte integer -> the general path is followed.
> 
> I don't see how.... The switch is for `sizeof(int)` and this is constant at runtime. Its NOT a switch for the data length passed in the OSSL_PARAM! So if OPENSSL_SMALL_FOOTPRINT is not defined it will never go to the general call. But even if it would, adding the check into the `general_get_[u]int()` functions (or before the call to them) would still be sufficient.

Oh yes it would, on any system where `int` isn't 32 or 64 bits.  These days, that's unusual, but...

**Comment by levitte** [comment]
BTW, a completely different solution could be this:

``` diff
diff --git a/crypto/bn/bn_lib.c b/crypto/bn/bn_lib.c
index 4fe6ce071a..cd902c2d57 100644
--- a/crypto/bn/bn_lib.c
+++ b/crypto/bn/bn_lib.c
@@ -446,6 +446,10 @@ static BIGNUM *bin2bn(const unsigned char *s, int len, BIGNUM *ret,
         return NULL;
     bn_check_top(ret);
 
+    /* If the input has no bits, the number is considered zero */
+    if (len == 0)
+        return ret;
+
     /*
      * The loop that does the work iterates from least to most
      * significant BIGNUM chunk, so we adapt parameters to transfer
```

**Comment by paulidale** [comment]
@ifranzki, you're correct :)

**Comment by levitte** [comment]
#20013 is now merged, so I rebased this

**Comment by ifranzki** [comment]
> BTW, a completely different solution could be this

I like that. This would also fix all uses of the xxx_2bn() functions that finally call bin2bn(). None of those seem to have dealt with the len = 0 case so far. Also it would allow OSSL_PARAMs to have data = NULL and data_size = 0 and thus would not cause any compatibility trouble where it is expected that this is treaded as a zero number.

Besides just returning ret in that case, I would make sure ret is actually set to zero before returning. If ret was passed in as already existing BIGNUM, it might not be zero, so just returning it might return a non-zero BIGNUM.

```
+    /* If the input has no bits, the number is considered zero */
+    if (len == 0) {
+        BN_clear(ret);
+        return ret;
+   }
+
```

**Comment by levitte** [comment]
The bin2bn fix is now submitted in #20033

**Pull request closed without merging**